### ISSUE TITLE
Dynamically determine loss function weights for binary classification

### ISSUE THREAD

**Issue by stephen-merrill** [dmlc/xgboost#2437]
I am (admittedly) new to XGBoost, but a thorough online search hasn't presented a solution to my problem.

I am attempting to build a binary classifier for fraud detection, and have data features like price, location, platform etc. The key to my classifier is that there are different type I and type II error costs here, and each observation has different costs because the costs are determined by the features in each observation.

I would like to turn these costs into weights and alter the 50/50 probability in a 0-1 decision into the appropriate probability as determined by the weights.

I see how to assign a custom weight that applies to the whole dataset, but I can't find a way to dynamically change the weights for each observation.

Does this amount to building a custom loss function? If so, any pointers? The examples I've found say they define the gradient and Hessian but I've had trouble seeing how.

**Comment by marugari** [comment]
That's hard.
https://github.com/dmlc/xgboost/issues/2255

However I don't believe that your objective is the following form:
`w1 * loss1(y[f1]) + w2 * loss2(y[f2]) + w3 * loss3(y[f3])`

**Comment by khotilov** [comment]
It's possible to assign individual weights to each observation in DMatrix. I'm not following what do you mean by needing "to _dynamically_ change the weights for each observation" and why. Boosting kind of does it for you.

**Comment by stephen-merrill** [comment]
It's not the observations themselves that I'm talking about weighting, it's the loss function for each observation. Here is an example:

Observation 1 is for a cheap product that has features that make it seem fraudulent. Hypothetical numbers here...

P(fraud) = 0.8
P(not fraud) = 0.2
Type I error cost = $50 (Because there is a high cost of offending a legitimate customer)
Type II error cost = $5 (Because the product doesn't cost very much)
THESE ERROR COSTS ARE DETERMINED BY THE FEATURE VALUES OF THE OBSERVATION, INDEPENDENT OF THE XGBOOST ALGORITHM!
Decision:
The loss from deciding fraud when it's legit is 50*0.2=10
The loss from deciding not fraud when it's fraud is 5*0.8=4
So I better say this observation is not fraud.

This is the type of behavior I need to have the XGBoost algorithm mimic. I know XGBoost can use the features from the data to classify observations as fraudulent or not, but I need to include this additional information about the costs of making these decisions, and these costs will change with each observation. For example, the price of each item will have huge effects on the type II error, and I may want to call an expensive transaction fraudulent even if the features don't offer much evidence toward fraud.

**Comment by khotilov** [comment]
Ok, so "dynamically" is perhaps just not a right word for what you are describing... I would suggest looking up example-dependent cost-sensitive classification - there should be some weighting methods (these individual weights will be used within the objective function), and also some ways to adjust individual thresholds on predicted risk in order to optimize total costs. Coming up with a custom loss that directly optimizes total cost and is also good behaving from the modeling perspective might be not an easy problem in this setting.

**Comment by stephen-merrill** [comment]
Thanks for your suggestion! I'll look into this.

**Comment by pommedeterresautee** [comment]
http://nbviewer.jupyter.org/format/slides/github/albahnsen/CostSensitiveClassification/blob/master/doc/tutorials/slides_edcs_fraud_detection.ipynb#/

**Comment by khotilov** [comment]
@pommedeterresautee Yes, that guy have recently done a lot in this field! Note that in his phd thesis https://github.com/albahnsen/phd-thesis on p66, where cost-sensitive logistic loss is introduced, there's also a mention that the function is not convex. That's what I meant by problematic behavior from the modeling perspective.

**Comment by geoHeil** [comment]
I recently read: https://link.springer.com/content/pdf/10.1007%2Fs10994-016-5572-x.pdf they were mostly comparing ada boost variants  and come to the conclusion that cost sensitive adding for boosting are not required - nonetheless still interesting read if you are interested in the topic of cost sensitive classification.

**Comment by khotilov** [comment]
@geoHeil that paper is about class-dependent costs, not example-dependent. And just like with logistic regression, class-dependent costs are not really an issue to worry too much about with gradient boosting. Some details, however, could have some effects, e.g., class-dependent weights might influence regularization, pruning, and things like that.

**Comment by binghesam** [comment]
by my experience, cost-sensitive is a little hard
